lib/kubernetes-deploy/kubernetes_resource/replica_set.rb

```diff
@@ -0,0 +1,116 @@
+# frozen_string_literal: true
+module KubernetesDeploy
+  class ReplicaSet < KubernetesResource
+    TIMEOUT = 5.minutes
+
+    def initialize(name:, namespace:, context:, file:, parent: nil, logger:)
+      @name = name
+      @namespace = namespace
+      @context = context
+      @file = file
+      @parent = parent
+      @logger = logger
+      @pods = []
+      @rollout_data = { "replicas" => 0 }
+    end
+
+    def sync
+      raw_json, _err, st = kubectl.run("get", type, @name, "--output=json")
+
+      if @found = st.success?
+        rs_data = JSON.parse(raw_json)
+        interpret_json_data(rs_data)
+      else # reset
+        @rollout_data = { "replicas" => 0 }
+        @status = nil
+        @pods = []
+      end
+    end
+
+    def interpret_json_data(rs_data)
+      pods_data = get_pods(rs_data)
+
+      pods_data.each do |pod_data|
+        pod = Pod.new(
+          name: pod_data["metadata"]["name"],
+          namespace: namespace,
+          context: context,
+          file: nil,
+          parent: "#{@name.capitalize} replica set",
+          logger: @logger
+        )
+        pod.deploy_started = @deploy_started
+        pod.interpret_json_data(pod_data)
+        @pods << pod
+      end
+
+      @rollout_data.merge!(rs_data["status"]
+        .slice("replicas", "availableReplicas", "readyReplicas"))
+      @status = @rollout_data.map { |st, num| "#{num} #{st.chop.pluralize(num)}" }.join(", ")
+    end
+
+    def deploy_succeeded?
+      @rollout_data["replicas"].to_i == @rollout_data["availableReplicas"].to_i &&
+      @rollout_data["replicas"].to_i == @rollout_data["readyReplicas"].to_i
+    end
+
+    def deploy_failed?
+      @pods.present? && @pods.all?(&:deploy_failed?)
```

Comment:
This (and `deploy_timed_out?`) still will basically never fail on account of the pods, since the Pod timeout is higher and the failure condition is super narrow (`@phase == "Failed"`). However, my next task is to make pod failures smarter / more aggressive, and at that point this should make a huge difference in how often our users hit the Deployment timeout.